tests/test_meta/test_estimatortransformer.py

```diff
@@ -85,3 +86,18 @@ def test_shape_multitarget(random_xy_dataset_multitarget):
         ]
     )
     assert pipeline.fit(X, y).transform(X).shape == (m, n)
+
+
+def test_kwargs(random_xy_dataset_clf):
+    """ Test if kwargs are properly passed to an underlying estimator. """
+    X, y = random_xy_dataset_clf
+    pipeline = EstimatorTransformer(LinearRegression())
+
+    # First sample is weighted twice as much
+    sample_weights = np.ones(shape=len(y))
+    sample_weights[0] = 2
+
+    # Check if fit method is explicitly called with sample weights
+    pipeline.fit = Mock()
+    pipeline.fit(X, y, model__sample_weight=sample_weights)
+    pipeline.fit.assert_called_with(X, y, model__sample_weight=sample_weights)
```

Comment:
@MBrouns Is this the correct way to use a `Mock` method for checking if the `sample_weight` argument is explcitily passed?

Reply:
almost, what you're checking now is whether the fit method of the `EstimatorTransformer` is called correctly (which it is because you do exactly that on line 102). We want to instead check if the `estimator` in the `Estimatortransformer` is called correctly


```
estimator = Mock()
pipeline = EstimatorTransformer(estimator)

...

pipeline.fit(X, y, model__sample_weight=sample_weights)
estimator.fit.assert_called_with(X, y, model__sample_weights=sample_weights)
```

Reply:
That makes sense, although struggling with implementation. `self.estimator_` is only introduced within the fit method as a clone of `self.estimator`. The adding of `Mock()` should therefore be created just after the clone, but before `EstimatorTransformer.estimator_.fit` in `EstimatorTransformer.fit`. Any idea how I can wrap `EstimatorTransformer.fit` so I can introducte `Mock()` just before the underlying estimator is being fitted or is there another way?

Reply:
Ah right, I forgot about the cloning! Interesting case. That essentially means we need to change how `clone` works, which I think we can do with `unittest.mock.patch`. 

```
from unittest.mock import patch


@patch('sklearn.clone')
def test_kwargs(patched_clone, random_xy_dataset_clf):
    """ Test if kwargs are properly passed to an underlying estimator. """
    X, y = random_xy_dataset_clf
    estimator = Mock()
    patched_clone.return_value = estimator
    
    pipeline = EstimatorTransformer(estimator)
    sample_weights = np.ones(shape=len(y))
    pipeline.fit(X, y, model__sample_weight=sample_weights)

    estimator.assert_called_with(X, y, model__sample_weight=sample_weights)
```

Reply:
Hmm, still some issues with that. `sklearn.clone` only works on scikit-learn objects so if it is `Mock()` it breaks (`TypeError: 'Mock' object is not iterable`).

If we set `estimator.fit = Mock()` and check `estimator.fit.assert_called_with` we get `AssertionError: expected call not found.` Not sure why that is, but it looks like it is not tracking the cloned object.

Solution I was thinking of:
```python
import sklearn
from unittest.mock import Mock
from unittest.mock import patch

@patch('sklearn.clone')
def test_kwargs(patched_clone, random_xy_dataset_clf):
    """ Test if kwargs are properly passed to an underlying estimator. """
    X, y = random_xy_dataset_clf
    estimator = LinearRegression()
    estimator.fit = Mock()
    patched_clone.return_value = estimator

    sample_weights = np.ones(shape=len(y))
    pipeline = EstimatorTransformer(estimator)
    pipeline.fit(X, y, sample_weight=sample_weights)

    estimator.fit.assert_called_with(X, y, sample_weight=sample_weights)
```

Reply:
To be honest I'm stuck on properly tracking this cloned object with `Mock`, @MBrouns. Do you have any suggestions on how to proceed?

Reply:
Okay that was indeed a bit trickier than I figured at first. The following works:

```

@patch('sklego.meta.estimator_transformer.clone')
def test_kwargs(patched_clone, random_xy_dataset_clf):
    """ Test if kwargs are properly passed to an underlying estimator. """
    X, y = random_xy_dataset_clf
    estimator = Mock()
    patched_clone.return_value = estimator

    sample_weights = np.ones(shape=len(y))
    pipeline = EstimatorTransformer(estimator)
    pipeline.fit(X, y, sample_weight=sample_weights)

    # We can't use `assert_called_with` becuase that compares by `==` which is ambiguous
    # on numpy arrays
    np.testing.assert_array_equal(
        X, estimator.fit.call_args.args[0]
    )
    np.testing.assert_array_equal(
        y, estimator.fit.call_args.args[1]
    )
    np.testing.assert_array_equal(
        sample_weights, estimator.fit.call_args.kwargs['sample_weight']
    )
```

The patch I suggested was wrong because in the `estimator_transformer.py` we import clone in to its namespace, rather than doing `import sklearn; sklearn.clone(...)` 

Also, the assert_called_with doesn't work on numpy arrays as it uses `==` by default, so I had to dig in to the `call_args`

Reply:
Interesting, thank you! This works like a charm! The tests are passing now. Looks like this is resolved.

Reply:
It seems that that functionality is implemented in 3.8 onwards indeed (https://bugs.python.org/issue21269). Could you try this:

```
  np.testing.assert_array_equal(
        X, estimator.fit.call_args[0][0]
    )
    np.testing.assert_array_equal(
        y, estimator.fit.call_args[0][1]
    )
    np.testing.assert_array_equal(
        sample_weights, estimator.fit.call_args[0]['sample_weight']
    )
```

Reply:
👍 Nice you could find the specific Python bugs issue. 

I pushed the code above after getting tests passing locally for Python3.7. I only had to change the last line, because `kwargs` are contained in the 2nd element of `call_args`. Like this:
```python
np.testing.assert_array_equal(
        sample_weights, estimator.fit.call_args[1]['sample_weight']
    )
```

Reply:
YES, that's it! Thank you, @MBrouns! 

Learned a lot from this.